File: synara/src/app/features/matrix-dto/receipt.ts
/**
 * Read-receipt DTO.
 */

import type { EventId, RoomId, UserId } from './ids';
import { hasForbiddenWireFields, isObject, optNumber, optString, reqString } from './parseUtil';

export const RECEIPT_TYPES = ['read', 'read_private', 'fully_read'] as const;
export type ReceiptType = typeof RECEIPT_TYPES[number];
const RECEIPT_TYPE_SET = new Set<string>(RECEIPT_TYPES);

export function isReceiptType(value: unknown): value is ReceiptType {
  return typeof value === 'string' && RECEIPT_TYPE_SET.has(value);
}

export type Receipt = {
  roomId: RoomId;
  eventId: EventId;
  userId: UserId;
  receiptType: ReceiptType;
  ts?: number;
  threadId?: EventId;
};

export function parseReceipt(value: unknown): Receipt | null {
  if (!isObject(value) || hasForbiddenWireFields(value)) return null;
  const roomId = reqString(value, 'roomId');
  const eventId = reqString(value, 'eventId');
  const userId = reqString(value, 'userId');
  const ts = optNumber(value, 'ts');
  const threadId = optString(value, 'threadId');
  if (
    roomId === null ||
    eventId === null ||
    userId === null ||
    ts === null ||
    threadId === null ||
    !isReceiptType(value.receiptType)
  ) {
    return null;
  }
  return {
    roomId,
    eventId,
    userId,
    receiptType: value.receiptType,
    ts,
    threadId,
  };
}

File: crates/synara-core/src/dto/receipt.rs
//! Read-receipt DTO.

use serde::{Deserialize, Serialize};

use super::ids::{EventId, RoomId, UserId};

/// Receipt kind (product enum).
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum ReceiptType {
    Read,
    ReadPrivate,
    FullyRead,
}

impl ReceiptType {
    pub fn as_str(self) -> &'static str {
        match self {
            Self::Read => "read",
            Self::ReadPrivate => "read_private",
            Self::FullyRead => "fully_read",
        }
    }
}

/// Single receipt projection.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Receipt {
    pub room_id: RoomId,
    pub event_id: EventId,
    pub user_id: UserId,
    pub receipt_type: ReceiptType,
    /// Receipt timestamp in milliseconds since Unix epoch.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub ts: Option<u64>,
    /// Thread root when the receipt is thread-scoped.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub thread_id: Option<EventId>,
}
